doc/website/blog/2022-03-31-ent-hooks-examples.md

```diff
@@ -0,0 +1,411 @@
+	require.Error(t, err)
+}
+```
+
+In this case, you can see that since "Yoni" and "Yolo" share the same 2 letters "Yo" the action fails.
+
+#### Recipe:
+
+First, register the hook:
+```go title="ent/schema/dog.go"
+// Hooks of the Dog.
+func (Dog) Hooks() []ent.Hook {
+	return []ent.Hook{		
+		hook.If(validateName,
+			hook.HasFields(dog.FieldOwnerID),
+		),
+	}
+}
+```
+
+Since we want our hook to call every time we change the owner (to make sure the name is matched against), I am using an
+[edge field](https://entgo.io/docs/schema-edges#edge-field). This provides the hooks a trigger when the edge's ID changes.
+
+The hook:
+```go title="ent/schema/dog.go"
+func validateName(next ent.Mutator) ent.Mutator {
+	return hook.DogFunc(func(ctx context.Context, m *gen.DogMutation) (ent.Value, error) {
+		owID, ok := m.OwnerID()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		owner, err := m.Client().User.Query().Where(user.ID(owID)).Only(ctx)
+		if err != nil {
+			return next.Mutate(ctx, m)
+		}
+		dn, ok := m.Name()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		if owner.Name[0:2] == dn[0:2] {
+			return nil, errors.New("invalid dog name")
+		}
+		return next.Mutate(ctx, m)
+	})
+}
+```
+A simple name validator and if we see a problem we cancel the mutation by returning an error before `next.Mutate(ctx,m)`.
+
+### Offloading Long Computations:
+
+In some cases, we have fields in our database that require long computation, we would like to make sure that this data is always updated once
+a change in the data requires it.
+
+A test for it will look like this:
+```go title="ent/schema/dog_test.go"
+func TestCacheHook(t *testing.T) {
+	ctx := context.Background()
+	cs := cache.NewSyncer()
+	c := enttest.Open(t, dialect.SQLite,
+		"file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1",
+		enttest.WithOptions(ent.CacheSyncer(cs)),
+	)
+	cs.Start(ctx, c)
+	cl := c.Cache.Create().SetWalks(-1).SaveX(ctx)
+	d := c.Dog.Create().SetName("Karashindo").SaveX(ctx)
+	u := c.User.Create().SetName("Yoni").
+		SetCache(cl).
+		AddPets(d).
+		SetConnectionString("mysql://root:pass@localhost:3306)").
+		SaveX(ctx)
+	c.Dog.UpdateOne(d).SetName("Fortuna").ExecX(ctx)
+	cs.Close()
+	cl = u.QueryCache().OnlyX(ctx)
+	require.True(t, cl.Walks > 0)
+}
+```
+This test is a bit more complicated, so let's review it line by line:
+First, I create a new `cache syncer` (remember we have a cache entity in our schema).
+
+This `Cache Syncer` can do heavy computations while running in a different context.
+We create an Ent client and inject our `Cache Syncer`, it is disabled by default.
+After that, we start it and provide it with the Ent client, the reason for that is to allow it to change our data (modify our cache entity).
+We create a few entities and update the dog's name, that will trigger a cache sync.
+Finally, we close the `Cache Syncer`, this will block until all pending hooks are completed.
+The assertion validates a value of "walks" larger than 0 (not -1).
```

Comment:
The example here has great potential, but it's really complex to understand and follow in one paragraph.